File: dipole_terry.py
from scipy.constants import g, mu_0, pi
import numpy as np
#np.seterr(all='raise') # 所有warning raise成error

class Magnet: # 固定的磁铁
    def __init__(self,pos,ma): # ma 是磁矩，和m区分
        self.pos = np.array(pos)
        self.ma = np.array(ma)

def dipole_force(m1, m2, r1, r2):
    R = r2 - r1
    R_mag = np.linalg.norm(R)
    if R_mag == 0:
        print(0)
        return np.zeros(3)
    R_hat = R / R_mag
    mu = mu_0 / (4 * pi)
    term1 = np.dot(m2, R_hat) * m1
    term2 = np.dot(m1, m2) * R_hat
    term3 = np.dot(m1, R_hat) * m2
    term4 = 5 * np.dot(m1, R_hat) * np.dot(m2, R_hat) * R_hat
    #print(term1,term2,term3,term4)
    try:
        return (3 * mu / R_mag**4) * (term1 + term2 + term3 - term4)
    except:
        print(r1,r2)

class Experiment: # 整个实验
    def __init__(self,x=0.0,y=0.0):
        ma = 1.0
        self.line_length = 0.4
        self.hanging_point = np.array([0.0,0.0,0.45])
        self.m = 0.02
        self.pos = np.array([0.0,0.0,0.0])
        self.ma = np.array([0.0,0.0,ma])
        self.pos[0] = x
        self.pos[1] = y
        self.dampingco = 0.01
        if x**2 + y**2 > self.line_length:
            raise Exception(f"line length {self.line_length} is not enough to get ({x},{y})")
        else:
            self.pos[2] = self.hanging_point[2] - np.sqrt(self.line_length**2 - self.pos[0]**2 - self.pos[1]**2)
        #print(self.pos)
        self.v = np.array([0.0,0.0,0.0])
        self.magnets = [Magnet([-0.1,0.0,0.0],[0.0,0.0,ma]),Magnet([0.1,0.0,0.0],[0.0,0.0,ma])]
        # self.magnets = [Magnet([0,0,0],[0,0,0])]
        self.dt = 0.01 #0.01 # dt
    def motion_with_tension(self):
        pos = self.pos
        vel = self.v
        m = self.m
        dmp = self.dampingco

        gravity = np.array([0, 0, -m * g])
        magnetic_force = 0
        for magnet in self.magnets:
            magnetic_force += dipole_force(magnet.ma, self.ma, magnet.pos, self.pos)
        damping = -dmp * vel

        F_ext = gravity + magnetic_force + damping

        radial = pos - self.hanging_point
        radial_unit = radial / np.linalg.norm(radial)

        v_squared = np.dot(vel, vel)
        radial_force = np.dot(F_ext, radial_unit)
        tension = m * v_squared / self.line_length + radial_force

        F_total = F_ext - tension * radial_unit
        if np.linalg.norm(F_total) > 1:
            pass
        acc = F_total / m

        return acc
    def simulation(
        self,
        x=0.0,
        y=0.0,
        max_time=70.0,
        terminate_acc=0.01,
        terminate_vel=0.02,
        respect_termination=True,
        v0=None,
    ):
        # Re-init state with optional custom initial velocity
        self.__init__(x, y)
        if v0 is not None:
            self.v = np.array(v0, dtype=float)
        t = 0
        T = []
        Pos = []
        while t < max_time:
            acc = self.motion_with_tension()
            self.v += acc*self.dt
            self.pos += self.v*self.dt

            radial = self.pos - self.hanging_point
            rhat = radial / np.linalg.norm(radial)
            self.pos = self.hanging_point + rhat * self.line_length
            # print(rhat)

            #    so v is purely tangential—no energy kick!
            self.v -= np.dot(self.v, rhat) * rhat

            t += self.dt
            T.append(t)
            Pos.append(self.pos.copy())
            if (
                respect_termination
                and np.linalg.norm(acc) <= terminate_acc
                and np.linalg.norm(self.v) <= terminate_vel
            ):
                #print(acc)
                #print(self.v)
                break
            #print(self.pos)
        return T,Pos

# e = Experiment()
# import matplotlib.pyplot as plt
# T,Pos = e.simulation(0.038,0.039)
# print(Pos[:10])
# from renderer import *
# render_trajectory(T,Pos,size=0.2,magnets=e.magnets,speed=1)
# render_d_t(T,Pos,0)
#plt.plot(T,[i[2] for i in Pos])
#plt.show()
# import matplotlib.pyplot as plt
# from renderer import *


def load_damping_data(path="damping1.txt"):
    """Read damping measurements as time, x, y lists."""
    data = np.genfromtxt(path, skip_header=2, delimiter="\t")
    if data.ndim == 1:
        data = data[np.newaxis, :]  # ensure 2D
    t_vals = data[:, 0].tolist()
    x_vals = data[:, 1].tolist()
    y_vals = data[:, 2].tolist()
    x_vals = [n+0.003 for n in x_vals]
    y_vals = [n+0.008 for n in y_vals]
    
    t_vals = t_vals[:4200]
    x_vals = x_vals[:4200]
    y_vals = y_vals[:4200]
    return t_vals, x_vals, y_vals

# damp_t, damp_x, damp_y = load_damping_data()

# sim_x = [p[0] for p in Pos]
# sim_y = [p[1] for p in Pos]

# fig, axes = plt.subplots(1, 2, figsize=(10, 4), sharex=False)

# # axes[0].plot(T, sim_x, label="Simulated trajectory", c="tab:blue", alpha = 0.5)
# # axes[0].plot(damp_t, damp_x, label="Empirical trajectory", c="tab:orange", alpha = 0.5)
# axes[0].plot(damp_t, damp_x, label="Empirical trajectory", c="tab:orange", alpha = 1)
# axes[0].set_xlabel("Time (s)")
# axes[0].set_ylabel("X (m)")
# axes[0].set_title("X vs Time")
# axes[0].legend()

# # axes[1].plot(T, sim_y, label="Simulated trajectory", c="tab:blue", alpha = 0.5)
# # axes[1].plot(damp_t, damp_y, label="Empirical trajectory", c="tab:orange", alpha = 0.5)
# axes[1].plot(damp_t, damp_y, label="Empirical trajectory", c="tab:orange", alpha = 1)
# axes[1].set_xlabel("Time (s)")
# axes[1].set_ylabel("Y (m)")
# axes[1].set_title("Y vs Time")
# axes[1].legend()

# fig.tight_layout()
# plt.show()

File: dipole_terry.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <cmath>

using namespace std;

const double g = 9.80665;
const double pi = 3.141592653589793;
const double mu_0 = 4*pi*1e-7;



struct Vector3 {
    double x, y, z;
    Vector3 operator+(const Vector3& o) const { return {x+o.x, y+o.y, z+o.z}; }
    Vector3 operator-(const Vector3& o) const { return {x-o.x, y-o.y, z-o.z}; }
    Vector3 operator*(double s)       const { return {x*s,   y*s,   z*s  }; }
    Vector3 operator/(double s)       const { return {x/s,   y/s,   z/s  }; }
    Vector3& operator+=(const Vector3& o) { x+=o.x; y+=o.y; z+=o.z; return *this; }
};
double dot(const Vector3& a, const Vector3& b) {
    return a.x*b.x + a.y*b.y + a.z*b.z;
}
double norm(const Vector3& v) {
    return sqrt(dot(v,v));
}
struct Magnet {
    Vector3 pos;
    Vector3 ma;
};
Vector3 dipole_force(const Vector3& ma1, const Vector3& ma2, const Vector3& r1, const Vector3& r2) {
    Vector3 R = r2 - r1;
    double Rmag = norm(R);
    if (Rmag == 0.0) return {0,0,0};
    Vector3 R_hat = R / Rmag;
    double term1 = dot(ma2, R_hat);
    double term2 = dot(ma1, ma2);
    double term3 = dot(ma1, R_hat);
    double term4 = dot(ma2, R_hat);

    Vector3 v1 = ma1 * term1;
    Vector3 v2 = R_hat * term2;
    Vector3 v3 = ma2 * term3;
    Vector3 v4 = R_hat * (5 * term3 * term4);

    return (v1 + v2 + v3 - v4) * (3 * mu_0 / (4*pi) / pow(Rmag, 4));
}

// parameter for experiment
Vector3 pos1_ = {-0.04,0,0}; // {-0.1,0,0}
Vector3 pos2_ = {0.04,0,0}; // {0.1,0,0}
Vector3 ma_ = {0,0,4.4}; // {0,0,3}
Vector3 ma1_ = {0,0,4.4}; // {0,0,3}
Vector3 ma2_ = {0,0,0}; // {0,0,3}
double m_ = 0.0075; // 0.02
Vector3 v_ = {0,0,0}; // {0,0,0}
double line_length_ = 0.27; // 0.4
Vector3 hanging_point_ = {0,0,0.315}; // {0,0,0.45}
double damping_ = 0.01; // 0.01

class Experiment {
public:
    double dt = 0.01; // 这是dt不要改！！！！！！！！
    double line_length;
    Vector3 hanging_point;
    double m;
    Vector3 ma;
    Vector3 pos;
    Vector3 v;
    double damping;
    vector<Magnet> magnets; // must be 2 elements!

    void init(double x, double y) {
        line_length = line_length_;        
        pos = {x, y, 0.0};
        if (x*x + y*y > line_length*line_length) {
            cout << x << " " << y << endl;
            throw runtime_error("line length too short for initial (x,y)");
        }
        double dz = sqrt(line_length*line_length - x*x - y*y);
        pos.z = hanging_point.z - dz;
        ma = ma_;
        m = m_;
        v = v_;
        damping = damping_;
        hanging_point = hanging_point_;
        magnets.clear();
        Magnet m1 = {pos1_, ma1_};
        magnets.emplace_back(m1);
        Magnet m2 = {pos2_, ma2_};
        magnets.emplace_back(m2);
    }

    Vector3 motion_with_tension() {
        Vector3 Fg = {0, 0, -m * g};
        Vector3 Fm = {0,0,0};
        for (auto& mag : magnets)
            Fm += dipole_force(mag.ma, ma, mag.pos, pos);
        // cout << Fm.x << endl;
        Vector3 Fd = v * -damping;
        Vector3 Fext = Fg + Fm + Fd;

        Vector3 radial = pos - hanging_point;
        double rnorm = norm(radial);
        Vector3 radial_hat = radial / rnorm;
        double v2 = dot(v,v);
        double radial_force = dot(Fext, radial_hat);
        double tension = m * v2 / line_length + radial_force;

        Vector3 Ftotal = Fext - radial_hat * tension;
        return Ftotal / m;
    }

    pair<vector<double>, vector<Vector3>>
    simulation(double x=0.0, double y=0.0,
               double max_time=80.0,
               double terminate_acc=0.01,
               double terminate_vel=0.02)
    {
        init(x, y);
        double t = 0.0;
        vector<double> T;
        vector<Vector3> Pos;

        while (t < max_time) {
            Vector3 acc = motion_with_tension();
            v += acc * dt;
            pos += v * dt;

            Vector3 rel = pos - hanging_point;
            Vector3 rhat = rel / norm(rel);
            pos = hanging_point + rhat * line_length;

            double radialComponent = dot(v,rhat);
            v = v - rhat * radialComponent;

            t += dt;
            T.push_back(t);
            Pos.push_back(pos);

            if (norm(acc) <= terminate_acc && norm(v) <= terminate_vel)
                break;
        }
        return {T, Pos};
    }
};

int main() {
    cout << "Hello";
    // ofstream f("D:\\Terry\\School\\Grade 11\\Phy\\ffy physics\\Magnetic Assist\\fractal_basin.txt");
    Experiment exp;
    pair<vector<double>, vector<Vector3>> result;
    double resolution = 500;
    double Sx = 0.15; // 0.2
    double Sy = 0.15; // 0.2
    // f << Sx << " " << Sy << endl;
    double dx = Sx*2/resolution;
    double dy = Sy*2/resolution;
    for (double y = -Sy; y < Sy; y += dy) {
        cout << "|";
        for (double x = -Sx; x < Sx; x += dx) {
            result = exp.simulation(x, y);
            auto T = result.first;
            auto Pos = result.second;
            auto p = Pos.back();
            double distance1 = norm(p - exp.magnets[0].pos);
            double distance2 = norm(p - exp.magnets[1].pos);
            Vector3 delta_height = {0, 0, exp.line_length};
            double distance_to_origin = norm(p - (exp.hanging_point - delta_height));
            double out;
            if (distance_to_origin <= 0.02) {
                out = 1;
            } else if (distance1 < distance2) {
                out = 0;
            } else {
                out = 2;
            }
            // f << out;
        }
        // f << endl;
    }
    // f.close();
    return 0;
}